algorithm.md

```diff
@@ -0,0 +1,174 @@
+potentially bidirectional. When we assign weights to edges, we give every edge
+a `toWeight` and `froWeight`. The `toWeight` configures how much cred flows
+from the `src` to the `dst`, and the `froWeight` configures how much cred flows
+from the `dst` to the `src`. This way, we don't need to add backwards edges to
+the graph to enable these flows. If an edge really should be uni-directional,
+we can just set one of the weights to 0.
+
+## Plugin Graphs
+
+A core goal for SourceCred is configurability and extensibility. That
+extensibility starts with the graph. The SourceCred core doesn't create any
+graphs itself; it just defines the semantics for graphs. Graphs are actually
+created by plugins. SourceCred currently includes two initial plugins: one for
+Git, another for GitHub. In the future, we imagine having plugins for other
+tools, like Twitter, StackOverflow, Google Docs, npm, and so forth.
+
+### Plugin Namespacing
+
+As mentioned above, we require that node and edge addresses be unique within a
+graph. That means that they need to be unique across plugins. To ensure this
+property, plugins are expected to define their own namespaces within address
+space, by making all of their node and edge addresses start with a unique
+two-part plugin prefix. The first element of the prefix should be the name of
+the owner of the plugin, and the second part should be the name of the plugin
+itself.
+
+For example, every node and edge address created by SourceCred's GitHub plugin
+starts with the prefix `["sourcecred", "github"]`.
+
+## Node and Edge Types
+
+Nodes and edges partition into meaningful "types" or "kinds" which share common
+semantics. For example, some kinds of nodes include issues, pull requests, and
+user accounts; kinds of edges include authors, references, and merges.
+
+To support this, SourceCred has a concept of [types]. Type membership is
+determined by a shared address prefix.
+
+[types]: https://github.com/sourcecred/sourcecred/blob/master/src/analysis/types.js
+
+For example, every node whose address starts with `["sourcecred", "git",
+"commit"]` is considered a member of the commit type defined by SourceCred's
+Git plugin.
+
+A node or edge may be a member of multiple types (and every node and edge is
+considered a member of the empty type).
+
+
+## Weights
+
+SourceCred allows the assignment of weights to nodes and edges. A weight is a
+non-negative floating point number, i.e. in the range `[0, Infinity)`. Nodes
+have a single weight, and edges have two weights, a `toWeight` and `froWeight`.
+At present, weights may only be set at the type level, but our intention is to
+allow arbitrarily customized weights that are modified by user provided
+*heuristics*.
+
+# Attribution
+
+Attribution is the process by which we assign a score ('cred') to every node in
+a contribution graph. It uses a PageRank-based algorithm, although in contrast
+to [canonical PageRank], there is no random teleportation vector.
```

Comment:
Might be worth mentioning that because we don’t have a teleportation
vector, the chain is not necessarily ergodic, so the algorithm can fail
to converge. Might also be worth mentioning that there’s no particular
reason that we don’t have the teleportation vector iirc.

Reply:
Done.